docs: add Popover documentation and live examples (#3587)
vaadin/docs: articles/components/popover/index.adoc

----

[source,tsx]
----
include::{root}/frontend/demo/component/popover/react/popover-positioning.tsx[render,tags=snippet,indent=0,group=React]
----
--

=== Target Gap

The distance between the popover and the target element can be customized by setting the following CSS properties on the Popover component:

- `--vaadin-popover-offset-top`
- `--vaadin-popover-offset-bottom`
- `--vaadin-popover-offset-start`
- `--vaadin-popover-offset-end`

=== Target Arrow

Popovers can render a wedge-shaped arrow tip pointing at the target element.

[.example]
--

[source,html]
----
include::{root}/frontend/demo/component/popover/popover-arrow.ts[render,tags=snippet,indent=0,group=Lit]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/popover/PopoverArrow.java[render,tags=snippet,indent=0,group=Flow]
----

[source,tsx]
----
include::{root}/frontend/demo/component/popover/react/popover-arrow.tsx[render,tags=snippet,indent=0,group=React]
----
--

== Dimensions

By default, the Popover’s size is determined by its contents, but an explicit width and height can be set on the Popover itself.

Contents that exceed the width of the popover will scroll.

> The maximum width of popovers is limited to the width of the viewport, minus a small margin that can be customized with CSS by overriding the *inset* property of the *vaadin-popover-overlay* element.

Reviewer: ```suggestion
The maximum width of popovers is limited to the width of the viewport, minus a small margin that can be customized with CSS by overriding the `inset` property of the `vaadin-popover-overlay` element.
```

== Modality

A modal Popover blocks the user from interacting with the rest of the user interface while open, automatically moves focus from the target element to the Popover, and traps keyboard focus within it.

When combined with an outside click closing trigger, modality prevents accidentally triggering other UI elements when clicking outside the Popover to close it.

By default, modal popovers do not render a modality curtain (or _backdrop_), but one can be enabled separately. A modality curtain can be useful for de-emphasizing the UI in the background and to give a visual indication that the rest of the UI is blocked from user interaction.

[.example]
--

[source,html]
----
include::{root}/frontend/demo/component/popover/popover-modal.ts[render,tags=snippet,indent=0,group=Lit]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/popover/PopoverModal.java[render,tags=snippet,indent=0,group=Flow]
----

[source,tsx]
----
include::{root}/frontend/demo/component/popover/react/popover-modal.tsx[render,tags=snippet,indent=0,group=React]
----
--

== Accessibility

> By default, the Popover overlay has the ARIA role *dialog*. This can be changed to another role to provide appropriate semantics for the type of content and interaction in the popover (see <<typical-use-cases, Typical Use Cases>> for examples).

Reviewer: ```suggestion
By default, the Popover overlay has the ARIA role `dialog`. This can be changed to another role to provide appropriate semantics for the type of content and interaction in the popover (see <<typical-use-cases, Typical Use Cases>> for examples).
```

> - *menu*: when the content is a list of actions or links

Reviewer: ```suggestion
- `menu:` when the content is a list of actions or links
```

> - *listbox*: when the content is a list form which you can select one or more items

Reviewer: ```suggestion
- `listbox:` when the content is a list form which you can select one or more items
```

> - *grid*: when the content is a tabular structure from which you select an item

Reviewer: ```suggestion
- `grid:` when the content is a tabular structure from which you select an item
```

> - *tree*: when the content is a hierarchical list

Reviewer: ```suggestion
- `tree:` when the content is a hierarchical list
```

Popover popover = new Popover();
popover.setAriaLabel("Label");
// OR
popover.setAriaLabelledby("label-element-id");
----

[source,html]
----
<source-info group="Lit"></source-info>
<vaadin-popover accessible-name="Label"></vaadin-popover>
<vaadin-popover accessible-name-ref="label-element-id"></vaadin-popover>

----

[source,tsx]
----
<source-info group="React"></source-info>
<Popover accessibleName="Label" />
<Popover accessibleNameRef="label-element-id" />
----

--

== Typical Use Cases

Here are a few examples of common use cases for the Popover component with recommended configurations.

=== Drop-Down Field

[.example]
--

[source,html]
----
include::{root}/frontend/demo/component/popover/popover-dropdown-field.ts[render,tags=snippet,indent=0,group=Lit]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/popover/PopoverDropdownField.java[render,tags=snippet,indent=0,group=Flow]
----

[source,tsx]
----
include::{root}/frontend/demo/component/popover/react/popover-dropdown-field.tsx[render,tags=snippet,indent=0,group=React]
----
--

- Opens on click, focus
- Closes on kbd:[Esc], blur, outside click, target click, and programmatically upon selection

> - Modal, no modality curtain (dropdown fields typically don’t have curtains)

Reviewer: ```suggestion
- Modal, no modality curtain (drop-down fields typically don’t have curtains)
```

> - Autofocused

Reviewer: ```suggestion
- Auto-focused
```

> - ARIA role=”dialog”

Reviewer: ```suggestion
- ARIA role `dialog`
```

> - aria-label “Select a date range”

Reviewer: ```suggestion
- ARIA label _“Select a date range”_
```

=== User Menu

[.example]
--

[source,html]
----
include::{root}/frontend/demo/component/popover/popover-user-menu.ts[render,tags=snippet,indent=0,group=Lit]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/popover/PopoverUserMenu.java[render,tags=snippet,indent=0,group=Flow]
----

[source,tsx]
----
include::{root}/frontend/demo/component/popover/react/popover-user-menu.tsx[render,tags=snippet,indent=0,group=React]
----
--

- Opens on click
- Closes on kbd:[Esc], outside click, target click
- Modal, no modality curtain (small popovers like this usually don’t need them)

> - ARIA role=”menu” (the overlay’s content is a menu; although there are non-list elements, they are not interactive, nor do they need to be announced by screen readers)

Reviewer: ```suggestion
- ARIA role `menu` (the overlay’s content is a menu; although there are non-list elements, they are not interactive, nor do they need to be announced by screen readers)
```

> - aria-label “User menu”

Reviewer: ```suggestion
- ARIA label _“User menu”_
```

Note: if the popover only needs to contain menu items, consider using a <<../menu-bar#,Menu Bar>> or <<../context-menu#,Context Menu>> instead.

=== Notification Panel

[.example]
--

[source,html]
----
include::{root}/frontend/demo/component/popover/popover-notification-panel.ts[render,tags=snippet,indent=0,group=Lit]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/popover/PopoverNotificationPanel.java[render,tags=snippet,indent=0,group=Flow]
----

[source,tsx]
----
include::{root}/frontend/demo/component/popover/react/popover-notification-panel.tsx[render,tags=snippet,indent=0,group=React]
----
--

- Opens on click
- Closes on kbd:[Esc], outside click, target click
- Modal, no modality curtain (small popovers like this usually don’t need them)

> - ARIA role=”dialog” (although mainly a list, there interactive non-list elements as well)

Reviewer: ```suggestion
- ARIA role `dialog` (although mainly a list, there interactive non-list elements as well)
```

> - aria-labelledby pointing to heading in overlay

Reviewer: ```suggestion
- ARIA labelled-by pointing to heading in overlay
```

- Arrow variant

=== Rich, Interactive Tooltip

[.example]
--

[source,html]
----
include::{root}/frontend/demo/component/popover/popover-interactive-tooltip.ts[render,tags=snippet,indent=0,group=Lit]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/popover/PopoverInteractiveTooltip.java[render,tags=snippet,indent=0,group=Flow]
----

[source,tsx]
----
include::{root}/frontend/demo/component/popover/react/popover-interactive-tooltip.tsx[render,tags=snippet,indent=0,group=React]
----
--

- Opens on hover, focus
- Closes on kbd:[Esc], outside click (as well as mouseout and blur)
- Non-modal
- No autofocus

> - ARIA role=”dialog” (“tooltip” would be invalid due to interactive contents)

Reviewer: ```suggestion
- ARIA role `dialog` (`tooltip` would be invalid due to interactive contents)
```

> - aria-labelledby pointing to heading in overlay

Reviewer: ```suggestion
- ARIA labelled-by pointing to heading in overlay
```

- Positioned above target
- Arrow variant

=== Anchored Modal Dialog

[.example]
--

[source,html]
----
include::{root}/frontend/demo/component/popover/popover-anchored-dialog.ts[render,tags=snippet,indent=0,group=Lit]
----

[source,java]
----
include::{root}/src/main/java/com/vaadin/demo/component/popover/PopoverAnchoredDialog.java[render,tags=snippet,indent=0,group=Flow]
----

[source,tsx]
----
include::{root}/frontend/demo/component/popover/react/popover-anchored-dialog.tsx[render,tags=snippet,indent=0,group=React]
----
--

- Opens on click
- Closes on kbd:[Esc], outside click
- Modal, with curtain

> - Autofocused

Reviewer: ```suggestion
- Auto-focused
```

> - ARIA role=”dialog”

Reviewer: ```suggestion
- ARIA role `dialog`
```

> - aria-labelledby pointing to heading in the overlay

Reviewer: ```suggestion
- ARIA labelled-by pointing to heading in the overlay
```
